File: 16-arithmetic-progression/task.erl
-module(task).

-export([ar_pr/2]).

ar_pr(Count, Values) ->
  ar_pr(Count, Values, []).

ar_pr(_, [], Result) ->
  lists:reverse(Result);
ar_pr(Count, [{V1, V2, V3}|ValuesTail], Result) ->
  Res1 = sum_pr(V3, V1, V2, 0),
  ar_pr(Count - 1, ValuesTail, [erlang:integer_to_binary(Res1)|Result]).

sum_pr(0, _, _, Sum) ->
  Sum;
sum_pr(Count, Begin, Step, Sum) ->
  if
    Count > 1 ->
      sum_pr(Count - 1, Begin, Step, Sum + (Begin + Step * (Count - 1)));
    true ->
      sum_pr(Count - 1, Begin, Step, Sum + Begin)
  end.

File: 16-arithmetic-progression/task.ex
defmodule Task do

  def ar_pr(count, values) do
    ar_pr(count, values, [])
  end

  defp ar_pr(_, [], result) do
    Enum.reverse(result)
  end

  defp ar_pr(count, [{v1, v2, v3}|tail_values], result) do
    res1 = sum_pr(v3, v1, v2, 0)
    ar_pr(count - 1, tail_values, [Integer.to_string(res1)|result])
  end

  defp sum_pr(0, _, _, sum) do
    sum
  end
  
  defp sum_pr(count, begin, step, sum) do
    if count > 1 do
      sum_pr(count - 1, begin, step, sum + (begin + step * (count - 1)))
    else
      sum_pr(count - 1, begin, step, sum + begin)
    end
  end

end
